File: src/orchestrator/contracts.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RiskLevel {
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum StepStatus {
    Pending,
    Approved,
    Executing,
    Success,
    Failed,
    Skipped,
    Blocked,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Route {
    MiddlewareApi,
    Midcli,
    Browser,
    Mcp,
}

impl Route {
    pub fn as_str(&self) -> &'static str {
        match self {
            Route::MiddlewareApi => "middleware_api",
            Route::Midcli => "midcli",
            Route::Browser => "browser",
            Route::Mcp => "mcp",
        }
    }
}

pub const ROUTE_PRIORITY: [Route; 4] = [
    Route::MiddlewareApi,
    Route::Midcli,
    Route::Browser,
    Route::Mcp,
];

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Action {
    pub domain: String,
    pub operation: String,
    pub resource: Value,
    #[serde(default)]
    pub args: Value,
    #[serde(default = "default_risk_level")]
    pub risk_level: RiskLevel,
    #[serde(default)]
    pub requires_approval: bool,
    #[serde(default)]
    pub dry_run: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecutionResult {
    pub task_id: String,
    pub step_id: String,
    pub executor_used: String,
    #[serde(default)]
    pub fallback_used: bool,
    pub status: StepStatus,
    #[serde(default)]
    pub duration_ms: u64,
    #[serde(default)]
    pub error_code: Option<String>,
    #[serde(default)]
    pub error_message: Option<String>,
    #[serde(default = "new_audit_ref")]
    pub audit_ref: String,
    #[serde(default)]
    pub result_payload: Value,
}

impl ExecutionResult {
    pub fn ok(&self) -> bool {
        self.status == StepStatus::Success
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskPlan {
    #[serde(default = "new_task_id")]
    pub task_id: String,
    #[serde(default)]
    pub goal: String,
    #[serde(default)]
    pub steps: Vec<Action>,
}

impl TaskPlan {
    pub fn from_json_str(input: &str) -> Result<Self, serde_json::Error> {
        let mut plan: TaskPlan = serde_json::from_str(input)?;
        for action in &mut plan.steps {
            action.normalize();
        }
        Ok(plan)
    }

    pub fn add(&mut self, mut action: Action) {
        action.normalize();
        self.steps.push(action);
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct TaskResult {
    pub task_id: String,
    pub results: Vec<ExecutionResult>,
}

impl TaskResult {
    pub fn new(task_id: String) -> Self {
        Self {
            task_id,
            results: Vec::new(),
        }
    }

    pub fn ok(&self) -> bool {
        self.results.iter().all(ExecutionResult::ok)
    }

    pub fn summary(&self) -> Value {
        let total_steps = self.results.len();
        let succeeded = self
            .results
            .iter()
            .filter(|r| r.status == StepStatus::Success)
            .count();
        let failed = self
            .results
            .iter()
            .filter(|r| r.status == StepStatus::Failed)
            .count();
        let blocked = self
            .results
            .iter()
            .filter(|r| r.status == StepStatus::Blocked)
            .count();
        serde_json::json!({
            "task_id": self.task_id,
            "total_steps": total_steps,
            "succeeded": succeeded,
            "failed": failed,
            "blocked": blocked,
            "ok": self.ok(),
        })
    }
}

impl Action {
    pub fn normalize(&mut self) {
        if matches!(self.risk_level, RiskLevel::High | RiskLevel::Critical) {
            self.requires_approval = true;
        }
    }
}

fn default_risk_level() -> RiskLevel {
    RiskLevel::Low
}

fn new_task_id() -> String {
    Uuid::new_v4().as_simple().to_string()
}

fn new_audit_ref() -> String {
    Uuid::new_v4().as_simple().to_string()[..12].to_string()
}

File: orchestrator/contracts.py
"""Canonical data contracts for the assistant runtime.

Action envelope and ExecutionResult follow the schema defined in
HarborBeacon-Middleware-Endpoint-Contract-v1.md and the SKILL.md
harboros-assistant-v2 spec.
"""
from __future__ import annotations

import uuid
from dataclasses import asdict, dataclass, field
from enum import Enum
from typing import Any


class RiskLevel(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class StepStatus(str, Enum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    EXECUTING = "EXECUTING"
    SUCCESS = "SUCCESS"
    FAILED = "FAILED"
    SKIPPED = "SKIPPED"
    BLOCKED = "BLOCKED"


class Route(str, Enum):
    MIDDLEWARE_API = "middleware_api"
    MIDCLI = "midcli"
    BROWSER = "browser"
    MCP = "mcp"


ROUTE_PRIORITY: list[Route] = [
    Route.MIDDLEWARE_API,
    Route.MIDCLI,
    Route.BROWSER,
    Route.MCP,
]

CONFIRMATION_REQUIRED_LEVELS: set[RiskLevel] = {RiskLevel.HIGH, RiskLevel.CRITICAL}


@dataclass
class Action:
    """Normalised action envelope."""
    domain: str
    operation: str
    resource: dict[str, Any]
    args: dict[str, Any] = field(default_factory=dict)
    risk_level: RiskLevel = RiskLevel.LOW
    requires_approval: bool = False
    dry_run: bool = False

    def __post_init__(self) -> None:
        if isinstance(self.risk_level, str):
            self.risk_level = RiskLevel(self.risk_level)
        if self.risk_level in CONFIRMATION_REQUIRED_LEVELS:
            self.requires_approval = True

    def to_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["risk_level"] = self.risk_level.value
        return d


@dataclass
class ExecutionResult:
    """Unified result from any executor."""
    task_id: str
    step_id: str
    executor_used: str
    fallback_used: bool = False
    status: StepStatus = StepStatus.PENDING
    duration_ms: int = 0
    error_code: str | None = None
    error_message: str | None = None
    audit_ref: str = field(default_factory=lambda: uuid.uuid4().hex[:12])
    result_payload: Any = None

    def __post_init__(self) -> None:
        if isinstance(self.status, str):
            self.status = StepStatus(self.status)

    @property
    def ok(self) -> bool:
        return self.status == StepStatus.SUCCESS

    def to_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["status"] = self.status.value
        return d


@dataclass
class TaskPlan:
    """A plan is a sequence of actions with dependencies."""
    task_id: str = field(default_factory=lambda: uuid.uuid4().hex)
    goal: str = ""
    steps: list[Action] = field(default_factory=list)

    def add(self, action: Action) -> str:
        step_id = f"s{len(self.steps) + 1}"
        self.steps.append(action)
        return step_id
